torch/nn/functional.py

```diff
@@ -439,16 +439,37 @@ def cross_entropy(input, target, weight=None, size_average=True):
     See :class:`torch.nn.CrossEntropyLoss` for details.
 
     Args:
-        input: Variable :math:`(N, C)` where `C = number of classes`
-        target: Variable :math:`(N)` where each value is `0 <= targets[i] <= C-1`
+        input: Variable :math:`(N, C)` or :math:`(N, C, H, W)`
+            where `C = number of classes`, `H = image height`,
+            and `W = image width`.
+        target: Variable :math:`(N)` or :math:`(N, H, W)`
+            where each value is `0 <= targets[i] <= C-1`
         weight (Variable, optional): a manual rescaling weight given to each
                 class. If given, has to be a Variable of size "nclasses"
         size_average (bool, optional): By default, the losses are averaged
                 over observations for each minibatch. However, if the field
                 sizeAverage is set to False, the losses are instead summed
                 for each minibatch.
     """
-    return nll_loss(log_softmax(input), target, weight, size_average)
+    ndim = input.ndim()
+    if ndim == 2:
+        return nll_loss(log_softmax(input), target, weight, size_average)
+    elif ndim() == 4:
+        N, C, H, W = input.size()
+        log_p = F.log_softmax(input)  # (N, C, H, W)
+        log_p = log_p.transpose(1, 2).transpose(2, 3).contiguous().view(-1, C)
+        log_p = log_p[target.view(N, H, W, 1).repeat(1, 1, 1, C) >= 0]
+        log_p = log_p.view(-1, C)  # (M, C) where M = (target >= 0).sum()
```

Comment:
To support Nd tensor, the code will be something like below,
but currently not sure how to handle `...` with PyTorch.
Should I create size variable like `size = target.size(); size.insert(0, 1);`?

```
N, C = input.size()[:2]
log_p = F.log_softmax(input)  # (N, C, ...)
log_p = log_p.transpose(0, 1).contiguous().view(C, -1)  # (C, X)
log_p = log_p[target.view(1, ...).repeat(C, 1, 1, ..., 1) >= 0]  # (M*C,)
log_p = log_p.view(-1, C)  # (M, C) where M = (target >= 0).sum()
```